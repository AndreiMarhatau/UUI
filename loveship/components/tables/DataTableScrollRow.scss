@use  '../../assets/styles/scss/prettifyScrollBar' as *;

.cell {
    box-sizing: border-box;
    padding: 0 12px;
}

:global(.uui-table-scroll-row-container) {
    @include prettifyCustomScroll();
}
