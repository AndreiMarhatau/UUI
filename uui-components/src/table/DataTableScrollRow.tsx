import * as React from 'react';
import cx from 'classnames';
import { ScrollManager, DataColumnProps, IHasCX } from '@epam/uui';
import { OverlayScrollbarsComponent } from 'overlayscrollbars-react';
import 'overlayscrollbars/css/OverlayScrollbars.css';
import { FlexCell } from '../layout/flexItems/FlexCell';
import { DataTableRowContainer } from './DataTableRowContainer';
import * as css from './DataTableScrollRow.scss';

export interface DataTableScrollRowProps extends IHasCX {
    columns?: DataColumnProps<any, any>[];
    scrollManager?: ScrollManager;
    cellClass?: string;
}

const uuiDataTableScrollRow = {
    uuiTableScrollRowContainer: 'uui-table-scroll-row-container',
    uuiTableScrollRow: 'uui-table-scroll-row',
    uuiTableScrollBar: 'uui-table-scroll-bar',
};

export class DataTableScrollRow extends React.Component<DataTableScrollRowProps, {}> {
    renderCell(column: DataColumnProps<any, any>) {
        return <FlexCell  cx={ [css.cellPlaceholder, this.props.cellClass] } { ...column } key={ column.key } />;
    }

    private clientWidth?: number;

    resizeObserver = new ResizeObserver(entries => {
        for (let entry of entries) {
            const contentRect = entry.contentRect;

            if (contentRect.width !== this.clientWidth) {
                this.forceUpdate();
            }
        }
    });

    componentWillUnmount() {
        this.resizeObserver.disconnect();
    }

    wrapScrollingSection(content: React.ReactNode, style: React.CSSProperties) {
        return (
            <OverlayScrollbarsComponent
                className={ uuiDataTableScrollRow.uuiTableScrollBar }
                style={ style }
                ref={ (scrollBars: OverlayScrollbarsComponent) => {
                    let node = scrollBars?.osInstance()?.getElements().viewport;
                    if (!node) return;
                    this.props.scrollManager && this.props.scrollManager.attachScrollNode(node);
                    this.resizeObserver && this.resizeObserver.observe(node);
                    this.clientWidth = node.clientWidth;
                } }
            >
                { content }
            </OverlayScrollbarsComponent>
        );
    }

    render() {
        return (
            <div className={ cx([uuiDataTableScrollRow.uuiTableScrollRowContainer, this.props.cx]) }>
                <DataTableRowContainer
                    cx={ uuiDataTableScrollRow.uuiTableScrollRow }
                    scrollManager={ this.props.scrollManager }
                    columns={ this.props.columns }
                    renderCell={ this.renderCell.bind(this) }
                    wrapScrollingSection={ this.wrapScrollingSection.bind(this) }
                />
            </div>
        );
    }
}