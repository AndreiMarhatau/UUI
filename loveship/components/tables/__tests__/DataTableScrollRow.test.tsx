import React from "react";
import renderer from "react-test-renderer";
import {testData} from "@epam/uui";
import {DataTableScrollRow} from "../DataTableScrollRow";

class ResizeObserverMock {
    observe = () => jest.fn();
    unobserve = () => jest.fn();
    disconnect = () => jest.fn();
}

global.ResizeObserver = ResizeObserverMock;

describe("DataTableScrollRow", () => {
    it("should be rendered correctly", () => {
        const tree = renderer
            .create(<DataTableScrollRow columns={ testData.dataColumns }/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});