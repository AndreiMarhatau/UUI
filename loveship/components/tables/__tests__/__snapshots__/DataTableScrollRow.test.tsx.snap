// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableScrollRow should be rendered correctly 1`] = `
<div
  className="uui-table-scroll-row-container cellClass"
>
  <div
    className="container uui-table-row-container uui-table-scroll-row container align-items-center"
  >
    <div
      className="uui-table-scroll-bar"
      style={
        Object {
          "flexGrow": 100,
          "flexShrink": 1,
          "height": "100%",
          "minWidth": 0,
          "overflow": "hidden",
          "position": "relative",
          "width": 0,
        }
      }
    >
      <div
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "bottom": 0,
            "left": 0,
            "marginBottom": 0,
            "marginRight": 0,
            "overflow": "scroll",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <div
          className="container align-items-center"
        >
          <div
            className="cellPlaceholder"
            onClick={[Function]}
            style={
              Object {
                "alignSelf": undefined,
                "flexBasis": "96px",
                "flexGrow": 0,
                "flexShrink": 0,
                "minWidth": 0,
                "textAlign": undefined,
              }
            }
          />
          <div
            className="cellPlaceholder"
            onClick={[Function]}
            style={
              Object {
                "alignSelf": undefined,
                "flexBasis": "96px",
                "flexGrow": 0,
                "flexShrink": 0,
                "minWidth": 0,
                "textAlign": undefined,
              }
            }
          />
          <div
            className="cellPlaceholder"
            onClick={[Function]}
            style={
              Object {
                "alignSelf": undefined,
                "flexBasis": "96px",
                "flexGrow": 0,
                "flexShrink": 0,
                "minWidth": 0,
                "textAlign": undefined,
              }
            }
          />
        </div>
      </div>
      <div
        style={
          Object {
            "borderRadius": 3,
            "bottom": 2,
            "display": "none",
            "height": 6,
            "left": 2,
            "position": "absolute",
            "right": 2,
          }
        }
      >
        <div
          className="uui-thumb"
        />
      </div>
      <div
        style={
          Object {
            "borderRadius": 3,
            "bottom": 2,
            "display": "none",
            "position": "absolute",
            "right": 2,
            "top": 2,
            "width": 6,
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,.2)",
              "borderRadius": "inherit",
              "cursor": "pointer",
              "display": "block",
              "position": "relative",
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
